EventFilter/HcalRawToDigi/plugins/alpaka/HcalDigisProducerPortable.cc

```diff
@@ -0,0 +1,200 @@
+    config_.maxChannelsF3HB = ps.getParameter<uint32_t>("maxChannelsF3HB");
+  }
+
+  void HcalDigisProducerPortable::produce(device::Event& event, device::EventSetup const& setup) {
+    const auto hbheDigis = event.getHandle(hbheDigiToken_);
+    const auto qie11Digis = event.getHandle(qie11DigiToken_);
+
+    //Get the number of samples in data from the first digi
+    auto const stride = HBHEDataFrame::MAXSAMPLES * 0.5 + 1;
+    auto const size = hbheDigis->size() * stride;  // number of channels * stride
+
+    // stack host memory in the queue
+    HostCollectionPhase0 hf5_(size, event.queue());
+
+    // device product
+    DeviceCollectionPhase0 df5_(size, alpaka::getDev(event.queue()));
+
+    // set SoA_Scalar;
+    hf5_.view().stride() = stride;
+    hf5_.view().size() = hbheDigis->size();
+
+    for (unsigned int i = 0; i < hbheDigis->size(); ++i) {
+      auto const& hbhe = (*hbheDigis)[i];
+      auto const id = hbhe.id().rawId();
+      auto const presamples = hbhe.presamples();
+      uint16_t header_word = (1 << 15) | (0x5 << 12) | (0 << 10) | ((hbhe.sample(0).capid() & 0x3) << 8);
+
+      auto hf5_vi = hf5_.view()[i];
+      hf5_vi.ids() = id;
+      hf5_vi.npresamples() = presamples;
+      hf5_vi.data()[0] = header_word;
+      //TODO:: get HEADER_WORDS/WORDS_PER_SAMPLE from DataFormat
+      for (unsigned int i = 0; i < hf5_.view().stride() - 1; i++) {
+        uint16_t s0 = (0 << 7) | (static_cast<uint8_t>(hbhe.sample(2 * i).adc()) & 0x7f);
+        uint16_t s1 = (0 << 7) | (static_cast<uint8_t>(hbhe.sample(2 * i + 1).adc()) & 0x7f);
+        uint16_t sample = (s1 << 8) | s0;
+        hf5_vi.data()[i + 1] = sample;
+      }
+    }
+    // copy hf5 to df5
+    alpaka::memcpy(event.queue(), df5_.buffer(), hf5_.const_buffer());
+
+    event.emplace(digisF5HBToken_, std::move(df5_));
+    //}
+    if (qie11Digis->empty()) {
+      event.emplace(digisF01HEToken_);
+      event.emplace(digisF3HBToken_);
+
+    } else {
+      auto size_f1 = 0;
+      auto size_f3 = 0;
+
+      // count the size of the SOA;
+      for (unsigned int i = 0; i < qie11Digis->size(); i++) {
+        auto const& digi = QIE11DataFrame{(*qie11Digis)[i]};
+
+        if (digi.flavor() == 0 or digi.flavor() == 1) {
+          if (digi.detid().subdetId() == HcalEndcap) {
+            size_f1++;
+          }
+        } else if (digi.flavor() == 3) {
+          if (digi.detid().subdetId() == HcalBarrel) {
+            size_f3++;
+          }
+        }
+      }
+      auto const nsamples = qie11Digis->samples();
+      auto const stride01 = nsamples * QIE11DataFrame::WORDS_PER_SAMPLE + QIE11DataFrame::HEADER_WORDS;
+
+      // stack host memory in the queue
+      HostCollectionPhase1 hf1_(size_f1, event.queue());
+      HostCollectionPhase1 hf3_(size_f3, event.queue());
+
+      // device product
+      DeviceCollectionPhase1 df1_(size_f1, alpaka::getDev(event.queue()));
+      DeviceCollectionPhase1 df3_(size_f3, alpaka::getDev(event.queue()));
```

Comment:
Also here use queue?
```suggestion
      DeviceCollectionPhase1 df1_(size_f1, event.queue());
      DeviceCollectionPhase1 df3_(size_f3, event.queue());
```

Reply:
Done.